grails-app/services/au/org/ala/userdetails/PasswordService.groovy

```diff
@@ -29,32 +35,237 @@ class PasswordService {
+     */
+    boolean checkUserPassword(User user, String password) {
+        if (!password || password.length() < 1) {
+            throw new IllegalArgumentException("The password must not be empty.")
+        }
+        if (user == null) {
+            throw new IllegalArgumentException("Must provide the user to compare a password.")
+        }
+
+        def existingPasswords = Password.findAllByUser(user)
+        def passwordType = getPasswordType()
+        def dateTimeNow = new Date()
+        def matchingPassword = existingPasswords.find { item ->
+            comparePasswords(password, item.password) &&
+                    item.type == passwordType &&
+                    (item.expiry == null || item.expiry > dateTimeNow) &&
+                    item.status == STATUS_CURRENT
+        }
+
+        if (matchingPassword) {
+            // delete any other passwords - there should only be one password for one user
+            existingPasswords.each {
+                if (it != matchingPassword) {
+                    it.delete()
+                }
+            }
+            return true
+        } else {
+            return false
+        }
+    }
+
+    /**
+     * Encode a password ready to store it.
+     * @param password The password to encode.
+     * @return The encoded password.
+     */
+    String encodePassword(String password) {
+        if (!password || password.length() < 1) {
+            throw new IllegalArgumentException("Must supply a password to be encoded.")
+        }
+
+        boolean isBcrypt = isBcrypt()
+        def encoder = isBcrypt ?
+                new BcryptPasswordEncoder(bcryptStrength) :
+                new LegacyPasswordEncoder(legacySalt, legacyAlgorithm, true)
+        def encodedPassword = encoder.encode(password)
+        return encodedPassword
+    }
+
+    /**
+     * Compare a plain-text password to an encoded password.
+     * @param plainPassword The plain-text password.
+     * @param hashedPassword The encoded password.
+     * @return True if the passwords match, otherwise false.
+     */
+    Boolean comparePasswords(String plainPassword, String hashedPassword) {
+        if (!plainPassword || plainPassword.length() < 1 || !hashedPassword || hashedPassword.length() < 1) {
+            throw new IllegalArgumentException("Must supply a plain text password and a hashed password to be compared.")
+        }
+
+        boolean isBcrypt = isBcrypt()
+        def encoder = isBcrypt ?
+                new BcryptPasswordEncoder(bcryptStrength) :
+                new LegacyPasswordEncoder(legacySalt, legacyAlgorithm, true)
+        def encodedPassword = encoder.matches(plainPassword, hashedPassword)
+        return encodedPassword
+    }
+
+    /**
+     * Validate a password against a password policy.
+     * @param username
+     * @param password
+     * @return
+     */
+    Map<String, Object> validatePassword(String username, String password) {
```

Comment:
Would you mind defining a class for the output of validatePassword, you can use groovy's `@Canonical` or `@Immutable` to take care of the Java boilerplate